contracts/ProtocolReserve/ProtocolShareReserve.sol

```diff
@@ -33,56 +94,172 @@ contract ProtocolShareReserve is Ownable2StepUpgradeable, ExponentialNoError, Re
 
-        __Ownable2Step_init();
-
-        protocolIncome = _protocolIncome;
-        riskFund = _riskFund;
+        CORE_POOL_COMPTROLLER = corePoolComptroller;
     }
 
     /**
      * @dev Pool registry setter.
      * @param _poolRegistry Address of the pool registry
      */
-    function setPoolRegistry(address _poolRegistry) external onlyOwner {
+    function setPoolRegistry(address _poolRegistry) external {
+        _checkAccessAllowed("setPoolRegistry(address)");
+
         require(_poolRegistry != address(0), "ProtocolShareReserve: Pool registry address invalid");
         address oldPoolRegistry = poolRegistry;
         poolRegistry = _poolRegistry;
-        emit PoolRegistryUpdated(oldPoolRegistry, _poolRegistry);
+        emit PoolRegistryUpdated(oldPoolRegistry, poolRegistry);
+    }
+
+    /**
+     * @dev Prime contract address setter.
+     * @param _prime Address of the prime contract
+     */
+    function setPrime(address _prime) external {
+        _checkAccessAllowed("setPrime(address)");
+
+        require(_prime != address(0), "ProtocolShareReserve: Prime address invalid");
+        address oldPrime = prime;
+        prime = _prime;
+        emit PrimeUpdated(oldPrime, prime);
+    }
+
+    /**
+     * @dev Fetches the list of prime markets and then accrues interest and
+     * releases the funds to prime for each market
+     */
+    function _accrueAndReleaseFundsToPrime() internal {
+        address[] memory markets = IPrime(prime).allMarkets();
+        for (uint i = 0; i < markets.length; i++) {
+            address market = markets[i];
+            IPrime(prime).accrueInterest(market);
+            _releaseFund(CORE_POOL_COMPTROLLER, IVToken(market).underlying());
+        }
+    }
+
+    /**
+     * @dev Fetches the list of prime markets and then accrues interest 
+     * to prime for each market
+     */
+    function _accruePrimeInterest() internal {
+        address[] memory markets = IPrime(prime).allMarkets();
+        for (uint i = 0; i < markets.length; i++) {
+            address market = markets[i];
+            IPrime(prime).accrueInterest(market);
+        }
+    }
+
+    /**
+     * @dev Add or update destination target based on destination address
+     * @param config configuration of the destination.
+     */
+    function addOrUpdateDistributionConfig(DistributionConfig memory config) external {
+        _checkAccessAllowed("addOrUpdateDistributionConfig(DistributionConfig)");
+        require(config.destination != address(0), "ProtocolShareReserve: Destination address invalid");
+
+        //we need to accrue and release funds to prime before updating the distribution configuration
+        //because prime relies on getUnreleasedFunds and it's return value may change after config update
+        _accrueAndReleaseFundsToPrime();
+
+        uint256 total = 0;
+        bool updated = false;
+        for (uint i = 0; i < distributionTargets.length; i++) {
```

Comment:
We can use a temporary variable to store `distributionTargets.length`, to save gas. Moreover, we can use `++i` instead of `i++`, also to save some gas